DSpace/dspace-angular
# Metadata edit patch error fix

## References
This PR closes #821 

## Description
This PR refactors the ItemMetadataComponent and ObjectUpdatesService to create a patch from the updates in the store and pass it on to the data-service's `patch` method instead of `update`. This allows us to create a better, more specific patch instead of comparing the old and new objects.

## Instructions for Reviewers
Changes made:
- Created a `MetadataPatchOperationService` (implementing new interface `PatchOperationService`), which is able to transform a `FieldUpdates` object from the store into an array of fast-json-patch operations. This happens in two major steps:
-- All updates are transformed into a `MetadataPatchOperation`. Each operation (add, remove and replace) has its own implementation of this wrapper object, each defining their own way to transform the information into an `Operation`.
-- These operations are iterated over and modified depending on previous operations and finally added to the patch to return. (for example, when two values of the same field are removed, the second remove operation should take the first into account, because it might alter its position)
- This service can be injected using its token: `METADATA_PATCH_OPERATION_SERVICE_TOKEN`
- When an object-updates-entry is initialised, you can now optionally pass an injection token with it. This way, we can store the service it should use for creating patches into the object-updates store. This is exactly what we'll be doing for our `ItemMetadataComponent`.
- Added a method `createPatch()` to the `ObjectUpdatesService`, which injects our new service using the token we stored and calls the `fieldUpdatesToPatchOperations()` method to create and return a patch
- Modified `ItemMetadataComponent` to use this new `createPatch()` method and pass the created patch on to the data-service's patch method. It now also displays error notifications if the patch fails.

How to test:
Visit the edit-metadata page for any item and perform a bunch of operations on its metadata like adding, removing and editing values, as well as combining multiple operations at once.
Also try to reproduce the issue from #821, which should now be fixed.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
